### PR TITLE
stopping retry while volume and node not found

### PR DESCRIPTION
**Pull request by arahamad**

### CODE REVIEW

**Comment by arahamad** [comment]
test result:

```
Select your choice
 1- Get volume details 
 2- Create snapshot 
 3- list snapshot 
 4- Create volume 
 5- Snapshot details 
 6- Snapshot Order 
 7- Create volume from snapshot
 8- Delete volume 
 9- Delete Snapshot 
 10- List all Snapshot 
 12- Authorize volume 
 13- Create VPC Volume 
 14- Create VPC Snapshot 
 15- Attach VPC volume 
 16- Detach VPC volume 
 17- Get volume by name 
 18- List volumes 
 19- Get volume Attachment 
Your choice?:
16
{"level":"info","ts":"2020-09-14T14:47:50.144+0530","caller":"utils/init_provider.go:125","msg":"Calling provider/utils/init_provider.go GenerateContextCredentials","RequestID":"49c3ae3f-4291-4e04-b426-4c235a8fe337"}
{"level":"info","ts":"2020-09-14T14:47:50.144+0530","caller":"utils/init_provider.go:140","msg":"Generating generateContextCredentials for ","RequestID":"49c3ae3f-4291-4e04-b426-4c235a8fe337","Provider ID":"iks-vpc-classic"}
{"level":"info","ts":"2020-09-14T14:47:50.144+0530","caller":"provider/provider.go:67","msg":"Entering IksVpcBlockProvider.OpenSession","RequestID":"49c3ae3f-4291-4e04-b426-4c235a8fe337"}
{"level":"info","ts":"2020-09-14T14:47:50.144+0530","caller":"provider/provider.go:72","msg":"Opening VPC block session","RequestID":"49c3ae3f-4291-4e04-b426-4c235a8fe337"}
{"level":"info","ts":"2020-09-14T14:47:50.144+0530","caller":"provider/provider.go:74","msg":"Its IKS dual session. Getttng IAM token for  VPC block session","RequestID":"49c3ae3f-4291-4e04-b426-4c235a8fe337"}
{"level":"info","ts":"2020-09-14T14:47:50.144+0530","caller":"iam/token_exchange_iks.go:106","msg":"In  tokenExchangeIKSRequest's sendTokenExchangeRequest()","RequestID":"49c3ae3f-4291-4e04-b426-4c235a8fe337"}
{"level":"info","ts":"2020-09-14T14:47:50.144+0530","caller":"iam/token_exchange_iks.go:126","msg":"Sending IAM token exchange request to container api server","RequestID":"49c3ae3f-4291-4e04-b426-4c235a8fe337"}
{"level":"info","ts":"2020-09-14T14:47:51.799+0530","caller":"provider/provider.go:207","msg":"Entering OpenSession","RequestID":"49c3ae3f-4291-4e04-b426-4c235a8fe337"}
{"level":"info","ts":"2020-09-14T14:47:51.799+0530","caller":"provider/provider.go:230","msg":"","RequestID":"49c3ae3f-4291-4e04-b426-4c235a8fe337","apiConfig.ContextID":"49c3ae3f-4291-4e04-b426-4c235a8fe337"}
{"level":"info","ts":"2020-09-14T14:47:51.799+0530","caller":"metrics/metrics.go:65","msg":"Time to complete","RequestID":"49c3ae3f-4291-4e04-b426-4c235a8fe337","OpenSession":0.000055076}
{"level":"info","ts":"2020-09-14T14:47:51.799+0530","caller":"provider/provider.go:87","msg":"Opening IKS block session","RequestID":"49c3ae3f-4291-4e04-b426-4c235a8fe337"}
{"level":"info","ts":"2020-09-14T14:47:51.799+0530","caller":"provider/provider.go:98","msg":"Its ISK dual session. Getttng IAM token for  IKS block session","RequestID":"49c3ae3f-4291-4e04-b426-4c235a8fe337"}
{"level":"info","ts":"2020-09-14T14:47:51.799+0530","caller":"iam/token_exchange_iks.go:106","msg":"In  tokenExchangeIKSRequest's sendTokenExchangeRequest()","RequestID":"49c3ae3f-4291-4e04-b426-4c235a8fe337"}
{"level":"info","ts":"2020-09-14T14:47:51.799+0530","caller":"iam/token_exchange_iks.go:126","msg":"Sending IAM token exchange request to container api server","RequestID":"49c3ae3f-4291-4e04-b426-4c235a8fe337"}
{"level":"info","ts":"2020-09-14T14:47:53.476+0530","caller":"provider/provider.go:207","msg":"Entering OpenSession","RequestID":"49c3ae3f-4291-4e04-b426-4c235a8fe337"}
{"level":"info","ts":"2020-09-14T14:47:53.476+0530","caller":"provider/provider.go:230","msg":"","RequestID":"49c3ae3f-4291-4e04-b426-4c235a8fe337","apiConfig.ContextID":"49c3ae3f-4291-4e04-b426-4c235a8fe337"}
{"level":"info","ts":"2020-09-14T14:47:53.476+0530","caller":"metrics/metrics.go:65","msg":"Time to complete","RequestID":"49c3ae3f-4291-4e04-b426-4c235a8fe337","OpenSession":0.000083968}
Enter the volume id: r134-ff7c81be-6a0a-4b07-9905-6158921006e0
Enter the instance id: test-bt53f0o20dilv5ntp3g0-openshift43-default-00000371
Enter the cluster id: bt53f0o20dilv5ntp3g0
{"level":"info","ts":"2020-09-14T14:48:24.543+0530","caller":"provider/detach_volume.go:30","msg":"Validating basic inputs for detach method...","RequestID":"49c3ae3f-4291-4e04-b426-4c235a8fe337","volumeAttachmentTemplate":{"volumeID":"r134-ff7c81be-6a0a-4b07-9905-6158921006e0","instanceID":"test-bt53f0o20dilv5ntp3g0-openshift43-default-00000371","vpcVolumeAttachment":{},"iksVolumeAttachment":{"clusterID":"bt53f0o20dilv5ntp3g0"}}}
{"level":"info","ts":"2020-09-14T14:48:24.543+0530","caller":"provider/detach_volume.go:41","msg":"Checking if volume is already attached ","RequestID":"49c3ae3f-4291-4e04-b426-4c235a8fe337"}
{"level":"info","ts":"2020-09-14T14:48:24.543+0530","caller":"provider/get_volume_attachment.go:26","msg":"Validating basic inputs for GetVolumeAttachment method...","RequestID":"49c3ae3f-4291-4e04-b426-4c235a8fe337","volumeAttachRequest":{"volumeID":"r134-ff7c81be-6a0a-4b07-9905-6158921006e0","instanceID":"test-bt53f0o20dilv5ntp3g0-openshift43-default-00000371","vpcVolumeAttachment":{},"iksVolumeAttachment":{"clusterID":"bt53f0o20dilv5ntp3g0"}}}
{"level":"info","ts":"2020-09-14T14:48:24.543+0530","caller":"provider/get_volume_attachment.go:79","msg":"Getting VolumeAttachmentList from VPC provider...","RequestID":"49c3ae3f-4291-4e04-b426-4c235a8fe337"}
{"level":"info","ts":"2020-09-14T14:48:24.543+0530","caller":"instances/iks_list_volume_attachments.go:39","msg":"Equivalent curl command and query parameters","RequestID":"49c3ae3f-4291-4e04-b426-4c235a8fe337","URL":"https://containers.test.cloud.ibm.com/v2/storage/vpc/getAttachmentsList?cluster=bt53f0o20dilv5ntp3g0\u0026generation=2\u0026version=2020-06-16\u0026worker=test-bt53f0o20dilv5ntp3g0-openshift43-default-00000371","volumeAttachmentTemplate":{"clusterID":"bt53f0o20dilv5ntp3g0","volume":{"id":"r134-ff7c81be-6a0a-4b07-9905-6158921006e0"}},"Operation":{"Name":"ListVolumeAttachment","Method":"GET","PathPattern":"v2/storage/vpc/getAttachmentsList"},"cluster":"bt53f0o20dilv5ntp3g0","worker":"test-bt53f0o20dilv5ntp3g0-openshift43-default-00000371"}
{"level":"error","ts":"2020-09-14T14:48:36.352+0530","caller":"instances/iks_list_volume_attachments.go:43","msg":"Error occured while getting volume attachments list","RequestID":"49c3ae3f-4291-4e04-b426-4c235a8fe337","error":"P4106: The worker node instance cannot be found."}
2020/09/14 14:48:36 TIME TAKEN BY FUNCTION IKS ListVolumeAttachments IS 11.80862287s
{"level":"info","ts":"2020-09-14T14:48:36.352+0530","caller":"provider/get_volume_attachment.go:40","msg":"Volume attachment response","RequestID":"49c3ae3f-4291-4e04-b426-4c235a8fe337","volumeAttachmentResponse":null,"error":"{Code:VolumeAttachFindFailed, Type:VolumeAttachFindFailed, Description:No volume attachment could be found for the specified volume ID 'r134-ff7c81be-6a0a-4b07-9905-6158921006e0' and instance ID 'test-bt53f0o20dilv5ntp3g0-openshift43-default-00000371'., BackendError:P4106: The worker node instance cannot be found., RC:400}"}
{"level":"info","ts":"2020-09-14T14:48:36.352+0530","caller":"provider/detach_volume.go:52","msg":"No volume attachment found for","RequestID":"49c3ae3f-4291-4e04-b426-4c235a8fe337","currentVolAttachment":null,"error":"{Code:VolumeAttachFindFailed, Type:VolumeAttachFindFailed, Description:No volume attachment could be found for the specified volume ID 'r134-ff7c81be-6a0a-4b07-9905-6158921006e0' and instance ID 'test-bt53f0o20dilv5ntp3g0-openshift43-default-00000371'., BackendError:P4106: The worker node instance cannot be found., RC:400}"}
{"level":"info","ts":"2020-09-14T14:48:36.352+0530","caller":"provider/detach_volume.go:64","msg":"Successfully detached volume from VPC provider","RequestID":"49c3ae3f-4291-4e04-b426-4c235a8fe337","resp":{"Status":"","StatusCode":200,"Proto":"","ProtoMajor":0,"ProtoMinor":0,"Header":null,"Body":null,"ContentLength":0,"TransferEncoding":null,"Close":false,"Uncompressed":false,"Trailer":null,"Request":null,"TLS":null}}
{"level":"info","ts":"2020-09-14T14:48:36.353+0530","caller":"metrics/metrics.go:65","msg":"Time to complete","RequestID":"49c3ae3f-4291-4e04-b426-4c235a8fe337","DetachVolume":11.809642747}
{"level":"info","ts":"2020-09-14T14:48:36.353+0530","caller":"provider/wait_for_detach_volume.go:28","msg":"Validating basic inputs for WaitForDetachVolume method...","RequestID":"49c3ae3f-4291-4e04-b426-4c235a8fe337","volumeAttachmentTemplate":{"volumeID":"r134-ff7c81be-6a0a-4b07-9905-6158921006e0","instanceID":"test-bt53f0o20dilv5ntp3g0-openshift43-default-00000371","vpcVolumeAttachment":{},"iksVolumeAttachment":{"clusterID":"bt53f0o20dilv5ntp3g0"}}}
{"level":"info","ts":"2020-09-14T14:48:36.353+0530","caller":"provider/get_volume_attachment.go:26","msg":"Validating basic inputs for GetVolumeAttachment method...","RequestID":"49c3ae3f-4291-4e04-b426-4c235a8fe337","volumeAttachRequest":{"volumeID":"r134-ff7c81be-6a0a-4b07-9905-6158921006e0","instanceID":"test-bt53f0o20dilv5ntp3g0-openshift43-default-00000371","vpcVolumeAttachment":{},"iksVolumeAttachment":{"clusterID":"bt53f0o20dilv5ntp3g0"}}}
{"level":"info","ts":"2020-09-14T14:48:36.353+0530","caller":"provider/get_volume_attachment.go:79","msg":"Getting VolumeAttachmentList from VPC provider...","RequestID":"49c3ae3f-4291-4e04-b426-4c235a8fe337"}
{"level":"info","ts":"2020-09-14T14:48:36.353+0530","caller":"instances/iks_list_volume_attachments.go:39","msg":"Equivalent curl command and query parameters","RequestID":"49c3ae3f-4291-4e04-b426-4c235a8fe337","URL":"https://containers.test.cloud.ibm.com/v2/storage/vpc/getAttachmentsList?cluster=bt53f0o20dilv5ntp3g0\u0026generation=2\u0026version=2020-06-16\u0026worker=test-bt53f0o20dilv5ntp3g0-openshift43-default-00000371","volumeAttachmentTemplate":{"clusterID":"bt53f0o20dilv5ntp3g0","volume":{"id":"r134-ff7c81be-6a0a-4b07-9905-6158921006e0"}},"Operation":{"Name":"ListVolumeAttachment","Method":"GET","PathPattern":"v2/storage/vpc/getAttachmentsList"},"cluster":"bt53f0o20dilv5ntp3g0","worker":"test-bt53f0o20dilv5ntp3g0-openshift43-default-00000371"}
{"level":"error","ts":"2020-09-14T14:48:43.777+0530","caller":"instances/iks_list_volume_attachments.go:43","msg":"Error occured while getting volume attachments list","RequestID":"49c3ae3f-4291-4e04-b426-4c235a8fe337","error":"P4106: The worker node instance cannot be found."}
2020/09/14 14:48:43 TIME TAKEN BY FUNCTION IKS ListVolumeAttachments IS 7.424188991s
{"level":"info","ts":"2020-09-14T14:48:43.777+0530","caller":"provider/get_volume_attachment.go:40","msg":"Volume attachment response","RequestID":"49c3ae3f-4291-4e04-b426-4c235a8fe337","volumeAttachmentResponse":null,"error":"{Code:VolumeAttachFindFailed, Type:VolumeAttachFindFailed, Description:No volume attachment could be found for the specified volume ID 'r134-ff7c81be-6a0a-4b07-9905-6158921006e0' and instance ID 'test-bt53f0o20dilv5ntp3g0-openshift43-default-00000371'., BackendError:P4106: The worker node instance cannot be found., RC:400}"}
{"level":"info","ts":"2020-09-14T14:48:43.777+0530","caller":"provider/wait_for_detach_volume.go:49","msg":"Volume detachment is complete","RequestID":"49c3ae3f-4291-4e04-b426-4c235a8fe337"}
{"level":"info","ts":"2020-09-14T14:48:43.777+0530","caller":"metrics/metrics.go:65","msg":"Time to complete","RequestID":"49c3ae3f-4291-4e04-b426-4c235a8fe337","WaitForDetachVolume":7.424341864}
Volume attachment &{ 200  0 0 map[] <nil> 0 [] false false map[] <nil> <nil>} <nil>


Select your choice
```

**Review by nkkashyap — COMMENTED** [review]
Looks good

**Review by prgavali — APPROVED** [review]
minot comment 
Can you rename `skipRetryForAttach` to something generic

**Review by nkkashyap — APPROVED** [review]

File: volume-providers/vpc/provider/detach_volume.go [hunk 1]
```diff
@@ -47,7 +47,7 @@ func (vpcs *VPCSession) DetachVolume(volumeAttachmentTemplate provider.VolumeAtt
 			volumeAttachment.ID = currentVolAttachment.VPCVolumeAttachment.ID
 			vpcs.Logger.Info("Detaching volume from VPC provider...")
 			response, err = vpcs.APIClientVolAttachMgr.DetachVolume(&volumeAttachment, vpcs.Logger)
-			return err, err == nil // Retry in case of all errors
+			return err, skipRetryForAttach(err, vpcs.Config.IsIKS)
```
[the remark this answers is not in the document]
Comment (arahamad) [reply volume-providers/vpc/provider/detach_volume.go]: error code which will be skip in this method are Obvious

**Review by prgavali — APPROVED** [review]
LGTM

**Review by prgavali — APPROVED** [review]

**Review by akgunjal — APPROVED** [review]
LGTM

**Pull request merged**